pyqtorch/api.py

```diff
@@ -96,12 +99,98 @@ def sample(
     )
 
 
+def analytical_expectation(
+    circuit: QuantumCircuit,
+    state: Tensor,
+    observable: Observable,
+    values: dict[str, Tensor] = dict(),
+    embedding: Embedding | None = None,
+) -> Tensor:
+    """Non sampled expectation value.
+
+    Given an initial state :math:`\\ket\\rangle`,
+    a quantum circuit :math:`U(\\theta)`,
+    the analytical expectation value with :math:`O` is defined as
+    :math:`\\langle\\bra U_{\\dag}(\\theta) |O| U(\\theta) \\ket\\rangle`
+
+    Args:
+        circuit (QuantumCircuit): Quantum circuit :math:`U(\\theta)`.
+        state (Tensor): Input state :math:`\\ket\\rangle`.
+        observable (Observable): Observable O.
+        values (dict[str, Tensor], optional): Parameter values for the observable if any.
+        embedding (Embedding | None, optional): An optional instance of `Embedding`.
+
+    Returns:
+        Tensor: Expectation value.
+    """
+    state = run(circuit, state, values, embedding=embedding)
+    return inner_prod(state, run(observable, state, values, embedding=embedding)).real
+
+
+def sampled_expectation(
+    circuit: QuantumCircuit,
+    state: Tensor,
+    observable: Observable,
+    values: dict[str, Tensor] = dict(),
+    embedding: Embedding | None = None,
+    n_shots: int = 1,
+) -> Tensor:
+    """Expectation value approximated via sampling.
+
+    Given an input state :math:`\\ket\\rangle`,
+    the expectation analytical value with :math:`O` is defined as
+    :math:`\\langle\\bra|O\\ket\\rangle`
+
+    Args:
+        state (Tensor): Input state :math:`\\ket\\rangle`.
+        observable (Observable): Observable O.
+        values (dict[str, Tensor], optional): Parameter values for the observable if any.
+        embedding (Embedding | None, optional): An optional instance of `Embedding`.
+        n_shots: (int, optional): Number of samples to compute expectation on.
+
+    Returns:
+        Tensor: Expectation value.
+    """
+    state = run(circuit, state, values, embedding=embedding)
+    n_qubits = circuit.n_qubits
+
+    # batchsize needs to be first dim for eigh
+    eigvals, eigvecs = torch.linalg.eigh(
+        observable.tensor(n_qubits=n_qubits, values=values).permute((2, 0, 1))
+    )
+    eigvals = eigvals.squeeze()
+    eigvec_state_prod = apply_operator(
+        state,
+        eigvecs.T.conj(),
+        tuple(range(n_qubits)),
+        n_qubits=circuit.n_qubits,
+    )
+    eigvec_state_prod = torch.flatten(eigvec_state_prod, start_dim=0, end_dim=-2).t()
+    probs = torch.abs(torch.pow(eigvec_state_prod, 2))
```

Comment:
Shouldn't that be the absolute squared instead ?

Reply:
Indeed. I got tricked by the same mistake in circuit.sample.